### ISSUE TITLE
Text boxes that use different font sizes

### ISSUE THREAD

**Issue by zeezee300** [bash0/cewe2pdf#185]
Hi,

I have a problem with text boxes that use different font sizes. For visual reasons, I have output the first letter of the first word with a larger font and in bold. However, the script seems to have a problem displaying the line spacing.
I have received a warning message:

```
10:22:11 - root - WARNING - A set of paragraphs would not fit inside its frame. Frame height is increased to prevent loss of text.
10:22:11 - root - WARNING -  Try widening the text box just slightly to avoid an unexpected word wrap, or increasing the height yourself
10:22:11 - root - WARNING -  Most recent paragraph text: <para autoLeading="max"><font name="Moderne Fraktur" size=14>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Test Text</font></para>
```

Can someone help me to solve this problem?

**Comment by AnEnglishmanInNorway** [comment]
Yes, I can see that technique might cause a problem, and I'm pretty sure
that the solution will not be simple, or even possible, for several
reasons. I haven't looked at the code to verify that these comments are
correct, but my first thoughts are these. We're obviously constrained by
what the reportlabs pdf creation package can do, so the first place to look
is there. We'd probably have to change the way we handle text quite
significantly, because we'd have to discover if the line height was going
to change - and we can't really tell that easily because it is reportlabs
that does the actual conversion of text strings into lines, in particular
it is reportlabs that starts a new line. Just in case anybody does feel
inspired to look at this, perhaps you could provide a little more
information?

   - I guess you're on Linux, since I don't have your Moderne Fraktur font
   on my Windows PC?
   - Could you add a screenshot of what your text box looks like in the
   CEWE editor?  And how it comes out in the pdf?

I'm also a bit curious as to how you added the non-breaking spaces in front
of your Test Text (showing as &nbsp; in the error message)?  That's a nice
idea to achieve simple spacing - though I'm uncertain if anybody has ever
tried it.

**Comment by zeezee300** [comment]
Thank you very much for the quick response.

I have created and attached an example:
This ist a screenshot from the Cewe application 
![picture cewe](https://github.com/user-attachments/assets/47eab6af-90c6-49a8-afad-08b027723f43)

and here the result from the cewe2pdf script:
![picture pdf](https://github.com/user-attachments/assets/65210400-ff56-435c-8704-bebf1784751c)


```
12:30:48 - root - WARNING - A set of paragraphs would not fit inside its frame. Frame height is increased to prevent loss of text.
12:30:48 - root - WARNING -  Try widening the text box just slightly to avoid an unexpected word wrap, or increasing the height yourself
12:30:48 - root - WARNING -  Most recent paragraph text: <para autoLeading="max"><font name="Moderne Fraktur" size=14>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Don Vamillo</font></para>
```

**Comment by AnEnglishmanInNorway** [comment]
Interesting work! I had a quick look using an example text with a large initial letter, like yours, and the problem is easy to reproduce. In the first of my trial screenshots, you can actually see that even CEWE doesn't get this exactly right. In the upper two text boxes, with the frames, you can see that both the first AND the second lines have a bigger line height!  But at least it's better than our pdf rendering in the second of the two screenshots!

---
![Issue185Cewe](https://github.com/user-attachments/assets/f626e8dc-916b-4545-9ef3-519cd8453f5d)
---
![Issue185Cewe2pdf](https://github.com/user-attachments/assets/b18fd434-cdb7-498d-82bd-1fdfaa6b7eef)
---

We create ReportLab paragraph objects from the HTML-like &lt;p&gt; paragraphs in the CEWE mcf file. Each CEWE &lt;p&gt; paragraph contains &lt;span&gt; objects which can vary the characteristics (e.g. font weight, slope, size, etc) of the continuous text in the paragraph, and we read and act on those.

The difficulty is that, as far as I know, a ReportLab paragraph must use the same leading (line height, basically) for the entire paragraph. Our "solution" is to choose the largest font size used in a paragraph to determine that line height for the paragraph. This is clearly not going to work for something where the font size varies in a paragraph, but it's the best we can do until somebody finds a better solution (either within the ReportLab framework, or by using a completely different pdf creation package)

I'm sure you are wondering how I did the third piece of text, bottom right in both images, where things seem to be OK. Basically I cheated. What looks like a single text box is in fact two boxes. The first one contains only the first line, with the initial large letter. The second box, placed just below it and carefully aligned, contains the second and subsequent lines of the text, all in the same smaller size. I don't know how much text you have to handle, but for just a few cases this technique might be a useful workaround. If you have a lot of material to handle, then I'm afraid cewe2pdf isn't going to be the tool for you, unless you want to make quite significant changes to your fork.

**Comment by AnEnglishmanInNorway** [comment]
What's the state on this, @zeezee300? Can we close it as having an acceptable workaround?
